File: py/tests/analysis/syntax/test_lexer.py
from collections.abc import Iterable
from dataclasses import dataclass

import pytest

from minsk.analysis.syntax import facts
from minsk.analysis.syntax.kind import SyntaxKind
from minsk.analysis.syntax.parser import SyntaxTree


@dataclass(frozen=True)
class SimpleToken:
    kind: SyntaxKind
    text: str


def get_tokens() -> Iterable[SimpleToken]:
    for kind in SyntaxKind:
        text = facts.get_text(kind)
        if text is not None:
            yield SimpleToken(kind, text)
    yield SimpleToken(SyntaxKind.NumberToken, "1")
    yield SimpleToken(SyntaxKind.NumberToken, "123")
    yield SimpleToken(SyntaxKind.IdentifierToken, "a")
    yield SimpleToken(SyntaxKind.IdentifierToken, "abc")


def get_separators() -> Iterable[SimpleToken]:
    yield SimpleToken(SyntaxKind.WhitespaceToken, " ")
    yield SimpleToken(SyntaxKind.WhitespaceToken, "  ")
    yield SimpleToken(SyntaxKind.WhitespaceToken, "\r")
    yield SimpleToken(SyntaxKind.WhitespaceToken, "\n")
    yield SimpleToken(SyntaxKind.WhitespaceToken, "\r\n")


def requires_separator(t1: SimpleToken, t2: SimpleToken) -> bool:
    t1_is_keyword = str(t1.kind).endswith("Keyword")
    t2_is_keyword = str(t2.kind).endswith("Keyword")

    if (t1.kind == SyntaxKind.IdentifierToken or t1_is_keyword) and (
        t2.kind == SyntaxKind.IdentifierToken or t2_is_keyword
    ):
        return True

    if t1.kind == SyntaxKind.NumberToken and t2.kind == SyntaxKind.NumberToken:
        return True

    if t1.kind in {SyntaxKind.BangToken, SyntaxKind.EqualsToken} and (
        t2.kind in {SyntaxKind.EqualsToken, SyntaxKind.EqualsEqualsToken}
    ):
        return True

    return False


def get_token_pairs() -> Iterable[tuple[SimpleToken, SimpleToken]]:
    for t1 in get_tokens():
        for t2 in get_tokens():
            if not requires_separator(t1, t2):
                yield t1, t2


def get_token_pairs_with_separator() -> (
    Iterable[tuple[SimpleToken, SimpleToken, SimpleToken]]
):
    for t1 in get_tokens():
        for t2 in get_tokens():
            if requires_separator(t1, t2):
                for sep in get_separators():
                    yield t1, sep, t2


def test_tests_all_tokens() -> None:
    tested_token_kinds = {
        t.kind for lst in (get_tokens(), get_separators()) for t in lst
    }

    untested_token_kinds = {
        k for k in SyntaxKind if str(k).endswith("Token") or str(k).endswith("Keyword")
    } - {
        SyntaxKind.BadToken,
        SyntaxKind.EndOfFileToken,
    }
    untested_token_kinds -= tested_token_kinds

    assert len(untested_token_kinds) == 0


@pytest.mark.parametrize("t", get_tokens())
def test_lexes_token(t: SimpleToken) -> None:
    tokens = SyntaxTree.parse_tokens(t.text)
    assert len(tokens) == 1
    assert tokens[0].kind == t.kind
    assert tokens[0].text == t.text


@pytest.mark.parametrize("t1,t2", get_token_pairs())
def test_lexes_token_pairs(t1: SimpleToken, t2: SimpleToken) -> None:
    text = t1.text + t2.text
    tokens = SyntaxTree.parse_tokens(text)
    assert len(tokens) == 2
    assert tokens[0].kind == t1.kind
    assert tokens[0].text == t1.text
    assert tokens[1].kind == t2.kind
    assert tokens[1].text == t2.text


@pytest.mark.parametrize("t1,sep,t2", get_token_pairs_with_separator())
def test_lexes_token_pairs_with_separator(
    t1: SimpleToken, sep: SimpleToken, t2: SimpleToken
) -> None:
    text = t1.text + sep.text + t2.text
    tokens = SyntaxTree.parse_tokens(text)
    assert len(tokens) == 3
    assert tokens[0].kind == t1.kind
    assert tokens[0].text == t1.text
    assert tokens[1].kind == sep.kind
    assert tokens[1].text == sep.text
    assert tokens[2].kind == t2.kind
    assert tokens[2].text == t2.text

File: nim/tests/testLexer.nim
import std/sequtils
import std/strformat
import std/strutils
import std/unittest

import minskpkg/codeAnalysis/syntax/[
  parser,
  syntaxKind,
  syntaxToken,
]

import helpers
import strutilsExt

type
  TestToken = object
    kind: SyntaxKind
    text: string

proc tt(kind: SyntaxKind, text: string): TestToken =
  result.kind = kind
  result.text = text

iterator getTokens(): TestToken =
  const tokens = [
    tt(SyntaxKind.IdentifierToken, "a"),
    tt(SyntaxKind.IdentifierToken, "abc"),
    tt(SyntaxKind.NumberToken, "1"),
    tt(SyntaxKind.NumberToken, "123"),
    tt(SyntaxKind.PlusToken, "+"),
    tt(SyntaxKind.MinusToken, "-"),
    tt(SyntaxKind.StarToken, "*"),
    tt(SyntaxKind.SlashToken, "/"),
    tt(SyntaxKind.BangToken, "!"),
    tt(SyntaxKind.EqualsToken, "="),
    tt(SyntaxKind.OpenParenthesisToken, "("),
    tt(SyntaxKind.CloseParenthesisToken, ")"),
    tt(SyntaxKind.AmpersandAmpersandToken, "&&"),
    tt(SyntaxKind.PipePipeToken, "||"),
    tt(SyntaxKind.EqualsEqualsToken, "=="),
    tt(SyntaxKind.BangEqualsToken, "!="),
    tt(SyntaxKind.FalseKeyword, "false"),
    tt(SyntaxKind.TrueKeyword, "true"),
  ]
  for t in tokens:
    yield t

iterator getSeparators(): TestToken =
  const tokens = [
    tt(SyntaxKind.WhitespaceToken, " "),
    tt(SyntaxKind.WhitespaceToken, "  "),
    tt(SyntaxKind.WhitespaceToken, "\r"),
    tt(SyntaxKind.WhitespaceToken, "\n"),
    tt(SyntaxKind.WhitespaceToken, "\r\n"),
  ]
  for t in tokens:
    yield t

proc requiresSeparator(t1Kind: SyntaxKind, t2Kind: SyntaxKind): bool =
  let t1IsKeyword = ($t1Kind).endsWith("Keyword")
  let t2IsKeyword = ($t2Kind).endsWith("Keyword")

  if (t1IsKeyword or t1Kind == SyntaxKind.IdentifierToken) and
      (t2IsKeyword or t2Kind == SyntaxKind.IdentifierToken):
    return true

  if t1Kind == SyntaxKind.NumberToken and t2Kind == SyntaxKind.NumberToken:
    return true

  if (
    t1Kind == SyntaxKind.WhitespaceToken and
    t2Kind == SyntaxKind.WhitespaceToken
  ):
    return true

  if t1Kind in {SyntaxKind.EqualsToken, SyntaxKind.BangToken} and
      t2Kind in {SyntaxKind.EqualsToken, SyntaxKind.EqualsEqualsToken}:
    return true

  return false

iterator getTokenPairs(): tuple[t1: TestToken, t2: TestToken] =
  for t1 in getTokens():
    for t2 in getTokens():
      if not requiresSeparator(t1.kind, t2.kind):
        yield (t1, t2)

iterator getTokenPairsWithSeparator(): tuple[
  t1: TestToken,
  sep: TestToken,
  t2: TestToken
] =
  for t1 in getTokens():
    for t2 in getTokens():
      if requiresSeparator(t1.kind, t2.kind):
        for sep in getSeparators():
          yield (t1, sep, t2)

suite "lexer":
  proc testLexerLexesToken(kind: SyntaxKind, text: string) =
    test fmt"lexer lexes token ({kind}, ""{text.readableEscape}"")":
      let token = parseTokens(text).assertSingle()
      assert token.kind == kind
      assert token.text == text

  proc testLexerLexesTokenPair(
    t1Kind: SyntaxKind,
    t1Text: string,
    t2Kind: SyntaxKind,
    t2Text: string
  ) =
    let text = t1Text & t2Text
    test (
      fmt"lexer lexes token pair ({t1Kind}, {t2Kind}," &
      fmt" ""{text.readableEscape}"")"
    ):
      let tokens = parseTokens(text)
      assert tokens.len == 2
      let t1 = tokens[0]
      let t2 = tokens[1]
      assert t1.kind == t1Kind
      assert t1.text == t1Text
      assert t2.kind == t2Kind
      assert t2.text == t2Text

  proc testLexerLexesTokenPairWithSeparator(
    t1Kind: SyntaxKind,
    t1Text: string,
    separatorKind: SyntaxKind,
    separatorText: string,
    t2Kind: SyntaxKind,
    t2Text: string
  ) =
    let text = t1Text & separatorText & t2Text
    test (
      fmt"lexer lexes token pair with separator ({t1Kind}, {t2Kind}," &
      fmt" ""{text.readableEscape}"")"
    ):
      let tokens = parseTokens(text)
      assert tokens.len == 3
      let t1 = tokens[0]
      let sep = tokens[1]
      let t2 = tokens[2]
      assert t1.kind == t1Kind
      assert t1.text == t1Text
      assert sep.kind == separatorKind
      assert sep.text == separatorText
      assert t2.kind == t2Kind
      assert t2.text == t2Text

  for t in getTokens().toSeq & getSeparators().toSeq:
    testLexerLexesToken(t.kind, t.text)

  for (t1, t2) in getTokenPairs():
    testLexerLexesTokenPair(
      t1.kind,
      t1.text,
      t2.kind,
      t2.text,
    )

  for (t1, sep, t2) in getTokenPairsWithSeparator():
    testLexerLexesTokenPairWithSeparator(
      t1.kind,
      t1.text,
      sep.kind,
      sep.text,
      t2.kind,
      t2.text,
    )
